uobtheatre/payments/payables.py

```diff
@@ -8,10 +12,41 @@ class Payable(metaclass=AbstractModelMeta):
     An model which can be paid for
     """
 
+    payments = GenericRelation(
+        Payment, object_id_field="pay_object_id", content_type_field="pay_object_type"
+    )
+
     @property
     @abc.abstractmethod
     def payment_reference_id(self):
         """
         The id of the payable object provided to payment providers.
         """
         raise NotImplementedError
+
+    @property
+    def provider_payment_value(self) -> int:
+        """The amount taken by the payment provider in paying for this object."""
+        return self.payments.filter(type=Payment.PaymentType.PURCHASE).aggregate(  # type: ignore
+            value=Sum("provider_fee")
+        )[
+            "value"
+        ]
+
+    @property
+    def app_payment_value(self) -> int:
+        """The amount taken by us in paying for this object."""
+        return self.payments.filter(type=Payment.PaymentType.PURCHASE).aggregate(  # type: ignore
+            value=Sum("app_fee") - Sum("provider_fee")
+        )[
+            "value"
+        ]
+
+    @property
+    def society_payment_value(self) -> int:
+        """The amount owed to the society for selling this object."""
+        return self.payments.filter(type=Payment.PaymentType.PURCHASE).aggregate(  # type: ignore
+            value=Sum("value") - Sum("app_fee")
+        )[
+            "value"
+        ]
```

Comment:
There is no logic in this file to calculate the amount to transfer to the society. (Description on line 47 is a bit incorrect in this sense)

I think we need another property like `society_transfer_value` or something which is basically `society_payment_value` but filtering for non-cash transactions? Also, what if `app_payment_value` comes out negative? I.e. our fee minus provider fee is negative? This difference would need to be deducted from a `society_transfer_value` in that instance.

We likely also want a property to calculate the total payments value (i.e. sum the value (total charged to customer)).

From the report point of view, the numbers we are interested for a production are: Total Sales, Total increase in square balance, Total fees for us, amount to be transferred to society.

Another point - how does this scale? On a production which might have over 300/400 payments, this looks like a lot of queries if we loop every performance, every booking and every payment.

Reply:
For now I would say:
- Performance of reports is a future problem
- If `app_payment_value` thats bad but for now lets just make sure that doesnt happen (we currently charge 5% theatre improvement levy and square charges us 1.9% so we should be good)